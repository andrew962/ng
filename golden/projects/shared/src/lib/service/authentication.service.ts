import { Injectable, OnDestroy, OnInit } from '@angular/core';
import { AngularFireAuth } from "@angular/fire/auth";
import { BehaviorSubject, Observable, of } from 'rxjs';
import firebase from 'firebase/app';
import { CurrentUser } from '../interfaces/currentUser.model';
import { Router } from '@angular/router';
import { AngularFireDatabase } from '@angular/fire/database';
import { UserPath } from '../enums/user.enum';

@Injectable({
  providedIn: 'root'
})
export class AuthenticationService implements OnDestroy {
  isLogged$: Observable<boolean>;
  private isLoggedSubject: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);
  currentUser$: Observable<CurrentUser>;
  private currentUserSubject: BehaviorSubject<CurrentUser> = new BehaviorSubject<CurrentUser>(null);
  superUser$: Observable<boolean>;
  private superUserSubject: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(null);
  constructor(
    private afAuth: AngularFireAuth,
    private router: Router,
    private db: AngularFireDatabase
  ) {
    this.isLogged$ = this.isLoggedSubject.asObservable();
    this.currentUser$ = this.currentUserSubject.asObservable();
    this.superUser$ = this.superUserSubject.asObservable();
  }

  ngOnDestroy(): void {

  }
  firebaseAuth() {
    return this.afAuth.signInWithPopup(new firebase.auth.GoogleAuthProvider()).then(r => {
      if (r != null) {
        this.db.database.ref(UserPath.DbUsersPath).child(r.user.uid).get().then(response => {
          this.superUserSubject.next(response.val().SuperUser);
        })
        this.setCurrentUser(r.user)
        return {
          isAuth: true
        }
      }
    })
  }

  firebaseLogout() {
    this.afAuth.signOut().then(() => {
      this.isLoggedSubject.next(false);
      this.currentUserSubject.next(null);
      this.router.navigateByUrl('/');
    })
  }

  setCurrentUser(r: firebase.User) {
    let currentUser: CurrentUser = {
      Uid: r.uid,
      UserName: r.displayName,
      UserEmail: r.email,
      PhotoURL: r.photoURL
    }
    this.isLoggedSubject.next(true);
    this.currentUserSubject.next(currentUser);
    sessionStorage.setItem('currentUser', JSON.stringify(currentUser));
  }

  checkState() {
    return this.afAuth.authState.subscribe({
      next: r => {
        if (r != null && r?.uid != null && r?.email != null) {
          this.db.database.ref(UserPath.DbUsersPath).child(r.uid).get().then(response => {
            this.superUserSubject.next(response.val().SuperUser);
          })
          this.isLoggedSubject.next(true);
          this.setCurrentUser(r)
        } else {
          this.isLoggedSubject.next(false);
        }
      }, error: err => this.isLoggedSubject.next(false)
    })
  }
}