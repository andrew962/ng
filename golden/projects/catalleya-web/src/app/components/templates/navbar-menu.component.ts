import { Component, OnInit } from '@angular/core';
import { CurrentUser } from 'projects/shared/src/lib/interfaces/currentUser.model';
import { AuthenticationService } from 'projects/shared/src/lib/service/authentication.service';
import { Observable } from 'rxjs';

@Component({
  selector: 'app-navbar-menu',
  templateUrl: './navbar-menu.component.html',
  styles: [
  ]
})
export class NavbarMenuComponent implements OnInit {
  isLogged$: Observable<boolean>;
  currentUser$: Observable<CurrentUser>
  constructor(
    private authenticationService: AuthenticationService
  ) {
    this.isLogged$ = this.authenticationService.isLogged$;
    this.currentUser$ = this.authenticationService.currentUser$;
  }

  ngOnInit(): void {}
  loginBtn() {
    this.authenticationService.firebaseAuth().then(response =>
      console.log(response))
  }
  logOut() {
    this.authenticationService.firebaseLogout()
  }

}
