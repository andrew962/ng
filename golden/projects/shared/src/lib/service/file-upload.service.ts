import { Injectable } from '@angular/core';
import { AngularFireStorage } from '@angular/fire/storage';
import { Observable, Subscription } from 'rxjs';
import { concatMap, finalize, tap } from 'rxjs/operators';
import { CurrentUser } from '../interfaces/currentUser.model';
import { AuthenticationService } from './authentication.service';
import { UserService } from './user.service';

@Injectable({
  providedIn: 'root'
})
export class FileUploadService {
  private uid: string;
  user: CurrentUser;
  authSubscription: Subscription;
  downloadURL: Observable<string>;
  constructor(
    private fStorage: AngularFireStorage,
    private auth: AuthenticationService,
    private userService: UserService
  ) {
    this.authSubscription = this.auth.currentUser$.subscribe(u => {
      this.user = u;
      this.uid = u?.Uid
    });
  }
  upload(file, fileName?: string, path: string = 'images') {
    let filePath: string = `${this.uid}/${path}/${fileName}`
    let fileRef = this.fStorage.ref(filePath);
    return this.fStorage.upload(filePath, file).snapshotChanges().pipe(concatMap(r => fileRef.getDownloadURL()));
  }
  getFile(filePath) {
    this.fStorage.ref(filePath).getDownloadURL().subscribe(r => {
      console.log(r)
    })
    // let filePath: string = `${this.uid}/${path}`
    // const fileRef = this.fStorage.ref(filePath);
    // this.downloadURL = fileRef.getDownloadURL();
    // return this.downloadURL.subscribe(r => {
    //   console.log(r)
    // })
  }
}
