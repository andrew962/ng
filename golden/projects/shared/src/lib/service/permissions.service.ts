import { Injectable, OnDestroy } from '@angular/core';
import { AngularFireAuth } from '@angular/fire/auth';
import { AngularFireDatabase } from '@angular/fire/database';
import { Subscription } from 'rxjs';
import { map } from 'rxjs/operators';
import { AuthenticationService } from './authentication.service';

@Injectable({
  providedIn: 'root'
})
export class PermissionsService implements OnDestroy {
  private uid: string;
  authSubscription: Subscription;
  private readonly rule = false
  constructor(
    private auth: AuthenticationService,
    private afDatabase: AngularFireDatabase,
  ) {
    this.authSubscription = this.auth.currentUser$.subscribe(u => this.uid = u?.Uid);
  }
  ngOnDestroy(): void {
    this.authSubscription.unsubscribe();
  }
  setPermissions() {

  }
}
