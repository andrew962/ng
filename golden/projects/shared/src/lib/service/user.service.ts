import { Injectable, OnDestroy } from '@angular/core';
import { AngularFireAuth } from '@angular/fire/auth';
import { AngularFireDatabase } from '@angular/fire/database';
import { Subscription } from 'rxjs';
import { UsersPath } from '../enums/user.enum';
import { CurrentUser } from '../interfaces/currentUser.model';
import { AuthenticationService } from './authentication.service';

@Injectable({
  providedIn: 'root'
})
export class UserService implements OnDestroy {
  private readonly rule = false
  private uid: string;
  user: CurrentUser;
  authSubscription: Subscription;
  constructor(
    private afAuth: AngularFireAuth,
    private db: AngularFireDatabase,
    private auth: AuthenticationService
  ) {
    this.authSubscription = this.auth.currentUser$.subscribe(u => {
      this.user = u;
      this.uid = u?.Uid
    });
  }
  ngOnDestroy(): void {
    this.authSubscription.unsubscribe();
  }

  createUser() {
    this.db.database.ref(UsersPath.DbUsersPath).child(this.uid).set({
      User: this.user,
      Permission: {
        Create: false,
        Update: false,
        Delete: false
      },
      Photos: []
    })
  }
  getUsers() {
    return this.db.database.ref(UsersPath.DbUsersPath).get()
  }
  updateUsers(data) {
    return this.db.database.ref(UsersPath.DbUsersPath).set(data);
  }
  updateUser() {
    this.db.database.ref(UsersPath.DbUsersPath).child(this.uid).update({
      Permission: {
        Create: true,
        Update: true,
        Delete: true
      }
    })
  }

}
