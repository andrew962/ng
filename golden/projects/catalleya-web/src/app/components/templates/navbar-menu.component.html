<nav class="navbar navbar-expand-md navbar-light navbar-bg justify-content-between">
    <a class="navbar-brand" [routerLink]="['']">Navbar</a>
    <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarNavAltMarkup"
        aria-controls="navbarTogglerDemo02" aria-expanded="false" aria-label="Toggle navigation">
        <span class="navbar-toggler-icon"></span>
    </button>
    <div class="collapse navbar-collapse" id="navbarNavAltMarkup">
        <ul class="navbar-nav mr-auto mt-2 mt-lg-0">
            <li class="nav-item" routerLinkActive="active">
                <a class="nav-link" [routerLink]="['']">Inicio</a>
            </li>
            <li class="nav-item" routerLinkActive="active">
                <a class="nav-link" [routerLink]="['/af-admin']">Admin</a>
            </li>
        </ul>
        <ng-container *ngIf="isLogged$ | async; else isNotLogged">
            <button class="btn btn-info btn-sm" (click)="logOut()">Cerrar sesion</button>
        </ng-container>
        <ng-template #isNotLogged>
            <button class="btn btn-info btn-sm" (click)="loginBtn()">Iniciar sesion</button>
        </ng-template>
    </div>
</nav>